include/dmlc/logging.h

```diff
@@ -66,11 +66,10 @@ inline void InitLogging(const char*) {
 
 class LogCheckError {
  public:
-  LogCheckError() : str(nullptr) {}
-  explicit LogCheckError(const std::string& str_) : str(new std::string(str_)) {}
-  ~LogCheckError() { if (str != nullptr) delete str; }
-  operator bool() {return str != nullptr; }
-  std::string* str;
+  LogCheckError() : str() {}
+  explicit LogCheckError(const std::string& str_) : str(str_) {}
+  operator bool() const { return !str.empty(); }
```

Comment:
I think an empty value is different from a missing value. Maybe we should use `std::unique_ptr<std::string>` to store the reference to the temporary string?

Reply:
In general yes, but here it's used for an error message, which should never be empty if present. In particular, it's only called as `LogCheckError("Error.")` or `LogCheckError(os.str())` with non-empty `os`.

Reply:
Anyway, this is now changed according to @tqchen's comment, probably `unique_ptr` would be as undesirable as my initial solution.